### PR TITLE
Export cohort to Saturn workspace

### PR DESCRIPTION
**Pull request by malathi13**
Exporting the filters to Saturn workspace
The filters are converted to SQL where clauses

### CODE REVIEW

File: api/data_explorer/__main__.py [hunk 1]
```diff
@@ -238,26 +217,38 @@ def _get_interval(max_field_value):
         return 1000000000000
 
 
-def _get_es_facets():
+def _process_facets():
     """Returns a dict from UI facet name to Elasticsearch facet object."""
+    """Returns a dict from UI facet name to UI facet description.
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: Update or delete. (This method doesn't return anything.)

File: api/data_explorer/controllers/export_url_controller.py [hunk 2]
```diff
@@ -60,6 +61,7 @@ def _check_preconditions():
 
     private_key_path = os.path.join(current_app.config['DATASET_CONFIG_DIR'],
                                     'private-key.json')
+    current_app.logger.info(private_key_path)
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Delete?

File: api/data_explorer/__main__.py [hunk 3]
```diff
@@ -268,24 +259,31 @@ def _get_es_facets():
...
-    return facets
+
+    app.app.logger.info('Elasticsearch facets: %s' % es_facets)
+    app.app.config['ELASTICSEARCH_FACETS'] = es_facets
+    app.app.logger.info('UI facets: %s' % ui_facets)
+    app.app.config['UI_FACETS'] = ui_facets
 
 
-def _get_table_names():
+def _process_bigquery_data():
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: nit _process_bigquery() or _process_bigquery_config()

File: api/data_explorer/controllers/export_url_controller.py [hunk 4]
```diff
@@ -69,7 +71,7 @@ def _check_preconditions():
         raise BadRequest(error_msg)
 
 
-def _get_entities_dict():
+def _get_entities_dict(cohortName, query):
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: [Python style](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#316-naming) is cohort_name

File: api/data_explorer/controllers/export_url_controller.py [hunk 5]
```diff
@@ -94,6 +96,14 @@ def _get_entities_dict():
                 'table_name': table_name
             }
         })
+    if query != "" and cohortName != "":
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: [Python style](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#2144-decision) is: `if query and cohort_name`:

File: api/data_explorer/controllers/export_url_controller.py [hunk 6]
```diff
@@ -132,11 +142,98 @@ def _create_signed_url(gcs_path):
...
+        high = high * 1000000000
+        return column + " >= " + str(low) + " AND " + column + " < " + str(high)
+
+    # For the ordinary ranges, there is no overlap between the ranges.
+    # So we include the upper end of the range. 
+    return column + " >= " + str(low) + " AND " + column + " <= " + str(high)
+
+
+def _get_filter_query(filter):
+    if filter is None or len(filter) == 0:
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: `if not filter or not len(filter):`

File: api/data_explorer/controllers/export_url_controller.py [hunk 7]
```diff
@@ -132,11 +142,98 @@ def _create_signed_url(gcs_path):
     return signed_url
 
 
+def _get_clause(column, type, value):
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: """Returns a single condition of a WHERE clause, eg "gender = female"."""

File: api/data_explorer/controllers/export_url_controller.py [hunk 8]
```diff
@@ -132,11 +142,98 @@ def _create_signed_url(gcs_path):
...
+    if type == 'text':
+        clause = column + ' = "' + value + '"'
+    elif type == 'boolean':
+        clause = column + ' =' + value
+    else:
+        clause = _get_range_clause(column, value)
+    return clause
+
+
+def _get_range_clause(column, value):
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: nit move to before _get_clause, since this is used by _get_clause

File: api/data_explorer/controllers/export_url_controller.py [hunk 9]
```diff
@@ -132,11 +142,98 @@ def _create_signed_url(gcs_path):
...
+    # So we include the upper end of the range. 
+    return column + " >= " + str(low) + " AND " + column + " <= " + str(high)
+
+
+def _get_filter_query(filter):
+    if filter is None or len(filter) == 0:
+        return ""
+    facets = current_app.config['UI_FACETS']
+    table_clauses = dict()
+    for f in filter:
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: FYI - After Bryan's stuff is in, we'll have to exclude files facets from the SQL query somehow. I'll talk to Brad about this.

File: api/data_explorer/controllers/export_url_controller.py [hunk 10]
```diff
@@ -132,11 +142,98 @@ def _create_signed_url(gcs_path):
...
+            table_clauses[table_name] = [
+                _get_clause(column, facet['type'], filter_value)
+            ]
+
+    table_selects = list()
+    primary_key = current_app.config['PRIMARY_KEY']
+    for table_name, clauses in table_clauses.iteritems():
+        table_select = "(SELECT %s FROM `%s` WHERE %s)"
+        where_clause = ""
+        for c in clauses:
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: nit [Style guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#316-naming) says not to abbreviate. Eg `c` should be `clause`

File: api/data_explorer/controllers/export_url_controller.py [hunk 11]
```diff
@@ -132,11 +142,98 @@ def _create_signed_url(gcs_path):
...
+    for table_select in table_selects:
+        table = "%s t%d" % (table_select, cnt)
+        join = " INNER JOIN %s ON t%d.%s = t%d.%s"
+        if cnt > 1:
+            query = query + join % (table, cnt - 1, primary_key, cnt,
+                                    primary_key)
+        else:
+            query = query + table
+        cnt = cnt + 1
+    current_app.logger.info("Final query %s" % query)
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Don't log since this is already logged in line 236

File: api/api.yaml [hunk 12]
```diff
@@ -57,6 +57,19 @@ paths:
         the url parameter set to this url. For each JSON file,
         https://rawls.dsde-prod.broadinstitute.org/#!/entities/create_entity
         may be called with the JSON as the POST body.
+      parameters:
```
[the remark this answers is not in the document]
Comment (malathi13) [reply api/api.yaml]: I'll do that in a separate PR, this is already big and rebase is difficult everytime I need to merge.

File: api/api.yaml [hunk 13]
```diff
@@ -57,6 +57,19 @@ paths:
...
+        - name: "cohortname"
```
[the remark this answers is not in the document]
Comment (malathi13) [reply api/api.yaml]: To clarify, the /exportUrl is currently a POST method. Are you saying we will make this a GET method? We can't have GET parameters for a POST method.

File: api/api.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (melissachang) [reply api/api.yaml]: Good idea

File: api/api.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (melissachang) [reply api/api.yaml]: Good point. I think technically query params are allowed for POST, but it's unusual. So body sounds good.

How about this:
```
      parameters:
        - name: "exportUrlRequest"
          in: "body"
          schema:
            type: object
            properties:
              cohortName:
                type: string
              filter:
                type: array
                # Default collectionFormat is csv. With csv, faceted search doesn't
                # work when facet value has a comma. So switch to pipes; pipe in facet
                # value is much more rare.
                collectionFormat: pipes
                items:
                  type: string
```

File: api/data_explorer/__main__.py [hunk 14]
```diff
@@ -238,26 +217,38 @@ def _get_interval(max_field_value):
         return 1000000000000
 
 
-def _get_es_facets():
-    """Returns a dict from UI facet name to Elasticsearch facet object."""
+def _process_facets():
+    """Process facets to store a dict from UI facet name to UI facet description
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: to UI facet description, Elasticsearch field name, and field type

File: api/data_explorer/__main__.py [hunk 15]
```diff
@@ -268,24 +259,31 @@ def _get_es_facets():
...
     Table names are fully qualified: <project id>.<dataset id>.<table name>
     If bigquery.json doesn't exist, this returns an empty list.
     """
     config_path = os.path.join(app.app.config['DATASET_CONFIG_DIR'],
                                'bigquery.json')
     table_names = []
+    primary_key = ""
     if os.path.isfile(config_path):
-        table_names = _parse_json_file(config_path)['table_names']
+        bigquery_data = _parse_json_file(config_path)
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: nit we usually call this `bigquery_config`, it's a bit more specific than bigquery_data

File: api/data_explorer/controllers/export_url_controller.py [hunk 16]
```diff
@@ -132,11 +141,93 @@ def _create_signed_url(gcs_path):
...
+            query = query + table
+        cnt = cnt + 1
+    return query
+
+
 def export_url_post():  # noqa: E501
     _check_preconditions()
-    entities = _get_entities_dict()
+    data = json.loads(request.data)
+    current_app.logger.info('Request data %s' % request.data)
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: nit let's be more specific "Export URL request data %s"

File: api/data_explorer/__main__.py [hunk 17]
```diff
@@ -238,26 +217,38 @@ def _get_interval(max_field_value):
...
-    facets = OrderedDict()
+    es_facets = OrderedDict()
+    ui_facets = OrderedDict()
 
     for facet_config in facets_config:
         field_name = facet_config['elasticsearch_field_name']
         field_type = _get_field_type(es, field_name)
         ui_facet_name = facet_config['ui_facet_name']
+
+        ui_facets[ui_facet_name] = {'name': field_name, 'type': field_type}
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: Sorry for not catching earlier -- can `name` be `elasticsearch_field_name`? `name` could be UI facet name. (Note even `field_name` is vague -- with upcoming variable exploration feature, it's not clear if `field_name` is Elasticsearch field name.)

File: api/data_explorer/__main__.py [hunk 18]
```diff
@@ -268,24 +259,31 @@ def _get_es_facets():
...
-            facets[ui_facet_name] = HistogramFacet(
+            es_facets[ui_facet_name] = HistogramFacet(
                 field=field_name, interval=_get_interval(max_field_value))
-    app.app.logger.info('Elasticsearch facets: %s' % facets)
-    return facets
+
+    app.app.logger.info('Elasticsearch facets: %s' % es_facets)
+    app.app.config['ELASTICSEARCH_FACETS'] = es_facets
+    app.app.logger.info('UI facets: %s' % ui_facets)
+    app.app.config['UI_FACETS'] = ui_facets
```
Comment (melissachang) [inline api/data_explorer/__main__.py]: opt - Elasticsearch field name isn't related to UI. Would be a bit cleaner to have just one facet dict with everything. This is optional, and can be in a different PR

File: api/data_explorer/controllers/export_url_controller.py [hunk 19]
```diff
@@ -132,11 +141,93 @@ def _create_signed_url(gcs_path):
...
+        high = int(high[:-1])
+        low = low * 1000000
+        high = high * 1000000
+    elif low.endswith('B'):
+        low = int(low[:-1])
+        high = int(high[:-1])
+        low = low * 1000000000
+        high = high * 1000000000
+
+    return column + " >= " + str(low) + " AND " + column + " < " + str(high)
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]:
```
# low is inclusive, high is exclusive
# See https://github.com/elastic/elasticsearch-dsl-py/blob/master/elasticsearch_dsl/faceted_search.py#L125
```

File: api/api.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (malathi13) [reply api/api.yaml]: Done.

File: api/api.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (malathi13) [reply api/api.yaml]: Done

File: api/data_explorer/__main__.py (hunk 14 above)
Comment (malathi13) [reply api/data_explorer/__main__.py]: Done

File: api/data_explorer/__main__.py (hunk 17 above)
Comment (malathi13) [reply api/data_explorer/__main__.py]: Done

File: api/data_explorer/__main__.py (hunk 15 above)
Comment (malathi13) [reply api/data_explorer/__main__.py]: Done

File: api/data_explorer/controllers/export_url_controller.py (hunk 19 above)
Comment (malathi13) [reply api/data_explorer/controllers/export_url_controller.py]: Done

File: api/data_explorer/controllers/export_url_controller.py (hunk 16 above)
Comment (malathi13) [reply api/data_explorer/controllers/export_url_controller.py]: Done

File: api/data_explorer/controllers/export_url_controller.py [hunk 20]
```diff
@@ -165,7 +165,7 @@ def _get_range_clause(column, value):
 
 
 def _get_clause(column, type, value):
-    """Returns a single condition of a WHERE clause, eg "gender = female"."""
+    """Returns a single condition of a WHERE clause, eg "gender = female, male"."""
```
Comment (melissachang) [inline api/data_explorer/controllers/export_url_controller.py]: Hmm, selecting both genders isn't realistic. Also, there should be ORs and ANDs. How about
```
"((age76 >= 20 AND age76 < 30) OR (age76 >= 30 AND age76 < 40))"
```
